guppylang/definition/struct.py

```diff
@@ -186,6 +193,33 @@ def check_instantiate(
         check_all_args(self.params, args, self.name, loc)
         return StructType(args, self)
 
+    @cached_property
+    def generated_methods(self) -> list[CustomFunctionDef]:
+        """Auto-generated methods for this struct."""
+
+        class ConstructorCompiler(CustomCallCompiler):
+            """Compiler for the `__new__` constructor method of a struct."""
+
+            def compile(self, args: list[OutPortV]) -> list[OutPortV]:
+                return [self.graph.add_make_tuple(args).out_port(0)]
+
+        constructor_sig = FunctionType(
+            inputs=[f.ty for f in self.fields],
+            output=StructType([p.to_bound(i) for i, p in enumerate(self.params)], self),
```

Comment:
How does this work, I assume the list is the `args` from `ParametrizedTypeBase`, then `self` is the `defn` from `StructType`. So in the default constructors for python dataclasses, you provide the fields for all of the parents, outer to inner?

Reply:
Yes, exactly. I could also use keyword args if it would be clearer?